src/serializer/ResidualHeapSerializer.js

```diff
@@ -624,30 +669,45 @@ export class ResidualHeapSerializer {
           functionValues,
           val
         );
+
         return { body, usedOnlyByResidualFunctions: true, description: "delay_initializer" };
       }
+    }
 
-      usedOnlyByAdditionalFunctions = numAdditionalFunctionReferences === functionValues.length;
+    if (additionalFunctionValue !== undefined) {
+      let additionalFunctionGenerator = this.additionalFunctionGenerators.get(additionalFunctionValue);
+      invariant(additionalFunctionGenerator !== undefined);
+      let body = this.activeGeneratorBodies.get(additionalFunctionGenerator);
+      invariant(body !== undefined);
+      return { body, usedOnlyByAdditionalFunctions: true };
     }
 
     let getBody = s => {
-      if (s === this.generator || s instanceof FunctionValue) {
-        return this.currentFunctionBody;
+      if (s === this.generator) {
+        return this.mainBody;
       } else {
         return this.activeGeneratorBodies.get(s);
       }
     };
 
+    generators = this._getReferencingGenerators(generators, functionValues).filter(
+      g => !this.additionalFunctionGeneratorsInverse.has(g)
+    );
+    invariant(generators.length > 0);
+
     // This value is referenced from more than one generator or function.
     // Let's find the body associated with their common ancestor.
-    let firstScope = scopes.values().next().value;
-    let commonAncestor = Array.from(scopes).reduce((x, y) => commonAncestorOf(x, y), firstScope);
+    let commonAncestor = Array.from(generators).reduce(
+      (x, y) => commonAncestorOf(x, y, this.getGeneratorParent),
+      generators[0]
```

Comment:
Chris, have a look at this in particular. I seem to remember that you recently changed it from generators[0] to scopes.values().next().value. Has the need of that change now gone away?

Reply:
I think so, also I believe Nikolai made that change.